src/tests/Microsoft.PowerFx.Core.Tests/LazyTypeCustomeDisplayNameProviderTests.cs

```diff
@@ -0,0 +1,120 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.Text;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Types;
+using Xunit;
+
+namespace Microsoft.PowerFx.Core.Tests
+{
+    public class LazyTypeCustomeDisplayNameProviderTests
+    {
+        private class LazyRecordType : RecordType
+        {
+            public override IEnumerable<string> FieldNames { get; }
+
+            public override bool TryGetFieldType(string name, out FormulaType type)
+            {
+                type = name switch
+                {
+                    "Num" => FormulaType.Number,
+                    "B" => FormulaType.Boolean,
+                    "Nested" => FormulaType.Unknown,
+                    "Inner" => FormulaType.Number,
+                    _ => FormulaType.Blank
+                };
+
+                return type != FormulaType.Blank;
+            }
+
+            public LazyRecordType()
+                : base(new CustomDisplayNameProvider())
+            {
+                FieldNames = new List<string>() { "Num", "B", "Nested", "Inner" };
+            }
+
+            public override bool Equals(object other)
+            {
+                return other is LazyRecordType;
+            }
+
+            public override int GetHashCode()
+            {
+                return 3;
+            }
+
+            private class CustomDisplayNameProvider : DisplayNameProvider
```

Comment:
This will depend on #583.

Note these methods are currently internal. 
We should land 583, and then this can implement the proper public surface